### PR TITLE
Make binaries and PDBs reproducible with ducible

### PR DESCRIPTION
**Pull request by lovelyyfiaaa**
Hiii ^^ 

This PR will solve https://github.com/wingtk/gvsbuild/issues/200 😊

This PR isn't readyy yeet! These are the things I have to do first:
- [x] run Ducible each step
- [x] have Ducible get installed like other tools (with [gvsbuild/tools.py](https://github.com/wingtk/gvsbuild/blob/main/gvsbuild/tools.py))
- [x] The __make_folder_reproducible() function actuallyy doesn't call ducible. Soo we need to fix this ^^
    - [x] ~~The above is now done, but everytine ducible is called, Windows will complain that it's missing ^^~~ This is now fixed!!
- [ ] Allow patching just some packages ^^

Edits:
- Omgg, wrong issue. Should have been #200

### CODE REVIEW

**Comment by fortressia** [comment]
Emm, I haven't check whether this code will run yet!.

**Comment by fortressia** [comment]
Part 2 of the todo isn't done yet! Soo CI will fail for now.

File: gvsbuild/utils/builder.py [hunk 1]
```diff
@@ -586,6 +586,28 @@ def build(self, projects):
...
+        proj = self.__project
+
+        for file in os.listdir(parent):
+            path = os.path.join(proj.pkg_dir, file)
+            if os.path.isdir(path):
+                self.__make_folder_reproducible(path)
+            else:
+                if file.endswith(".exe") or file.endswith(".dll"):
+                    noextension_name = file[:-4]
+                    pdb_name = os.path.join(parent, noextension_name, ".pdb")
```
Comment (cas--) [inline gvsbuild/utils/builder.py:604]: Just a passing suggestion that you could simplify things here with pathlib and use subprocess

```python
from pathlib import Path
import subprocess

paths = Path(parent).glob('**/*') 
files = [p for p in paths if p.is_file() and p.suffix in ['.exe', '.dll']]
for file in files:
    pdb = file.with_suffix('.pdb')
    if not pdb.exists():
        pdb = ''
    image = Path(self.__project.pkg_dir) / file
    subprocess.run(f"ducible {image} {pdb}", shell=True)
```

Comment (fortressia) [reply gvsbuild/utils/builder.py:604]: Aww, thanks!

**Comment by fortressia** [comment]
Emmm, can anyone help me? How do I get ducible installed with the tools things in Gvsbuild?

File: gvsbuild/utils/builder.py (hunk 1 above)
Comment (fortressia) [reply gvsbuild/utils/builder.py]: Mergedd!! <3

**Comment by fortressia** [comment]
Great!!! I now know what to do ^^
```
Building project python ()
Traceback (most recent call last):
  File "E:\projects\gvsbuild\gvsbuild\utils\builder.py", line 536, in build
    if self.__build_one(p):
  File "E:\projects\gvsbuild\gvsbuild\utils\builder.py", line 676, in __build_one
    proj.prepare_build_dir()
  File "E:\projects\gvsbuild\gvsbuild\utils\base_project.py", line 390, in prepare_build_dir
    self.unpack()
  File "E:\projects\gvsbuild\gvsbuild\utils\base_project.py", line 399, in unpack
    raise NotImplementedError("unpack")
NotImplementedError: unpack
Error: ducible build failed
```

**Comment by fortressia** [comment]
So the next step is going to be implementing that `unpack` function.

**Comment by fortressia** [comment]
A lovely screenshot of File Explorer showing ducible installed to the `C:\gtk-build\tools\ducible-1.2.2`

Greatt!! Now that we have this installed. We can move on to making sure that the executable actually gets executed (because it doesn't seem so in my testing) and allowing reproducing partially!

**Comment by fortressia** [comment]
Ahh, I see! So the problem here is that the globs don't match the files. I wonder thoo.

image


image

**Comment by fortressia** [comment]
Great! We just need to fix the path issue now ^^

Windows complaining that ducible is not in the path

**Comment by lovelyyfiaaa** [comment]
OMGG why didn't I realized thiss? https://mesonbuild.com/Reproducible-builds.html

Soo this PR is placeboo?

**Pull request closed without merging**